openml/runs/trace.py

```diff
@@ -26,56 +30,348 @@ def get_selected_iteration(self, fold, repeat):
+    def to_filesystem(self, file_path):
+        """Serialize the trace object to the filesystem.
+
+        Serialize the trace object as an arff.
+
+        Parameters
+        ----------
+        file_path: str
+            File path where the trace arff will be stored.
+        """
+
+        trace_arff = arff.dumps(self.trace_to_arff())
+        with open(os.path.join(file_path, 'trace.arff'), 'w') as f:
+            f.write(trace_arff)
+
+    def trace_to_arff(self):
+        """Generate the arff dictionary for uploading predictions to the server.
+
+        Uses the trace object to generate an arff dictionary representation.
+
+        Returns
+        -------
+        arff_dict : dict
+            Dictionary representation of the ARFF file that will be uploaded.
+            Contains information about the optimization trace.
+        """
+        if self.trace_iterations is None:
+            raise ValueError("trace_iterations missing from the trace object")
+
+        # attributes that will be in trace arff
+        trace_attributes = [
+            ('repeat', 'NUMERIC'),
+            ('fold', 'NUMERIC'),
+            ('iteration', 'NUMERIC'),
+            ('evaluation', 'NUMERIC'),
+            ('selected', ['true', 'false']),
+        ]
+        trace_attributes.extend([
+            ('parameter_' + parameter, 'STRING') for parameter in
+            next(iter(self.trace_iterations.values())).get_parameters()
+        ])
+
+        arff_dict = OrderedDict()
+        data = []
+        for trace_iteration in self.trace_iterations.values():
+            tit_list = []
+            for attr, _ in trace_attributes:
+                if attr.startswith('parameter_'):
+                    attr = attr[len('parameter_'):]
+                    value = trace_iteration.get_parameters()[attr]
+                else:
+                    value = getattr(trace_iteration, attr)
+                if attr == 'selected':
+                    if value:
+                        tit_list.append('true')
+                    else:
+                        tit_list.append('false')
+                else:
+                    tit_list.append(value)
+            data.append(tit_list)
+
+        arff_dict['attributes'] = trace_attributes
+        arff_dict['data'] = data
+        arff_dict['relation'] = "Trace"
+        return arff_dict
+
+    @classmethod
+    def trace_from_arff(cls, arff_obj):
+        """Generate trace from arff trace.
+
+        Creates a trace file from arff object (for example, generated by a
+        local run).
+
+        Parameters
+        ----------
+        arff_obj : dict
+            LIAC arff obj, dict containing attributes, relation, data.
+
+        Returns
+        -------
+        OpenMLRunTrace
+        """
+        trace = OrderedDict()
+        # flag if setup string is in attributes
+        attribute_idx = {
+            att[0]: idx for idx, att in enumerate(arff_obj['attributes'])
+        }
+        required_attributes = [
```

Comment:
From here on, the code seems highly similar to the static function that you pulled in the generate function. Unless there is any good reason, why not put the duplicate code in a (static) functions?